multicluster/controllers/multicluster/gateway_controller.go

```diff
@@ -0,0 +1,331 @@
+
+func (r *GatewayReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	klog.V(2).InfoS("Reconciling Gateway", "gateway", req.NamespacedName)
+	requeue, commonArea, err := r.getCommonArea()
+	if requeue {
+		return ctrl.Result{Requeue: true}, nil
+	}
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	r.leaderNamespace = commonArea.GetNamespace()
+	err = r.getServiceCIDR(ctx)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	resExportName := r.localClusterID + "-clusterinfo"
+	resExportNamespaced := types.NamespacedName{
+		Name:      resExportName,
+		Namespace: r.leaderNamespace,
+	}
+	resExport := &mcsv1alpha1.ResourceExport{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      resExportName,
+			Namespace: r.leaderNamespace,
+		},
+	}
+
+	createOrUpdate := func(gwIP string) error {
+		existingResExport := &mcsv1alpha1.ResourceExport{}
+		if err := commonArea.Get(ctx, resExportNamespaced, existingResExport); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return err
+			}
+			if err = r.createorUpdateResourceExport(ctx, req, commonArea, existingResExport, gwIP, true); err != nil {
+				return err
+			}
+			return nil
+		}
+		if err = r.createorUpdateResourceExport(ctx, req, commonArea, existingResExport, gwIP, false); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	gwn := &mcsv1alpha1.Gateway{}
+	if err := r.Client.Get(ctx, req.NamespacedName, gwn); err != nil {
+		if apierrors.IsNotFound(err) {
+			gwNodeInfo, err := r.getLastCreatedGateway()
+			if err != nil {
+				klog.ErrorS(err, "Failed to get Gateways")
+				return ctrl.Result{}, err
+			}
+			if gwNodeInfo == nil {
+				// When the last Gateway is deleted, we will remove the ClusterInfo kind of ResourceExport
+				if err := commonArea.Delete(ctx, resExport, &client.DeleteOptions{}); err != nil {
+					return ctrl.Result{}, client.IgnoreNotFound(err)
+				}
+				return ctrl.Result{}, nil
+			}
+			// When there are still Gateways exist, we should create or update existing ResourceExport
+			// with the latest Gateway in remaining Gateways.
+			if err := createOrUpdate(gwNodeInfo.GatewayIP); err != nil {
+				return ctrl.Result{}, err
+			}
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+	if err := createOrUpdate(gwn.GatewayIP); err != nil {
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+// getCommonArea checks ClusterSet initialization status. When ClusterSet is initialized successfully by member
+// ClusterSet controller, remoteCommonAreaManager will be built and include a connected CommonArea.
+func (r *GatewayReconciler) getCommonArea() (requeue bool, remoteCommonArea commonarea.RemoteCommonArea, err error) {
+	defer r.mutex.RUnlock()
+	r.mutex.RLock()
+	if *r.remoteCommonAreaManager == nil {
+		klog.InfoS("ClusterSet has not been initialized properly, no remote cluster manager")
+		return true, nil, nil
```

Comment:
And why we need to retry here?

Reply:
We need to make sure the gateway events can be handled after the ClusterSet is initialized. otherwise, the event will be gone and not handled unless we restart the controller.